File: agents/android/app/src/main/java/today/ionity/csi/Protocol.kt
package today.ionity.csi

import kotlinx.serialization.Serializable

/** JSON shapes mirroring @ionity/metrification (validated server-side by zod). */

@Serializable
data class AccessPoint(
    val bssid: String,
    val ssid: String,
    val freqMhz: Double,
    val bandwidthMhz: Double? = null,
    val rssi: Double? = null,
)

@Serializable
data class RssiSample(val rssi: Double, val ts: Long)

@Serializable
data class CsiFrame(
    val ts: Long,
    val amplitude: List<Double>,
    val phase: List<Double>,
    val rssi: Double? = null,
)

@Serializable
data class Measurement(
    val rssi: List<RssiSample>,
    val csi: List<CsiFrame>,
    val ftmDistanceMm: Double? = null,
    val ftmStdDevMm: Double? = null,
)

@Serializable
data class CalibrationLabel(
    val axis: String,
    val valueMeters: Double,
    val direction: String? = null,
    val repetition: Int,
)

@Serializable
data class Capture(
    val deviceId: String,
    val agent: String = "android",
    val ap: AccessPoint,
    val startTs: Long,
    val endTs: Long,
    val measurement: Measurement,
    val label: CalibrationLabel? = null,
    val sessionId: String? = null,
)

@Serializable
data class StatusMessage(
    val deviceId: String,
    val agent: String = "android",
    val ts: Long,
    val state: String,
    val scan: List<AccessPoint>? = null,
    val message: String? = null,
)

object Topics {
    fun capture(id: String) = "ionity/$id/capture"
    fun status(id: String) = "ionity/$id/status"
}

File: agents/linux-nexmon/ionity_agent/protocol.py
"""Build protocol messages (must match @ionity/metrification zod schemas)."""
from __future__ import annotations

TOPIC_ROOT = "ionity"


def topic_capture(device_id: str) -> str:
    return f"{TOPIC_ROOT}/{device_id}/capture"


def topic_status(device_id: str) -> str:
    return f"{TOPIC_ROOT}/{device_id}/status"


def build_capture(
    device_id: str,
    ap: dict,
    start_ts: int,
    end_ts: int,
    rssi_samples: list[dict],
    csi_frames: list[dict],
    ftm_distance_mm: float | None = None,
    label: dict | None = None,
    session_id: str | None = None,
) -> dict:
    measurement: dict = {"rssi": rssi_samples, "csi": csi_frames}
    if ftm_distance_mm is not None:
        measurement["ftmDistanceMm"] = ftm_distance_mm
    capture: dict = {
        "deviceId": device_id,
        "agent": "linux-nexmon",
        "ap": ap,
        "startTs": start_ts,
        "endTs": end_ts,
        "measurement": measurement,
    }
    if label is not None:
        capture["label"] = label
    if session_id is not None:
        capture["sessionId"] = session_id
    return capture


def build_status(device_id: str, state: str, ts: int, scan: list[dict] | None = None, message: str | None = None) -> dict:
    status: dict = {"deviceId": device_id, "agent": "linux-nexmon", "ts": ts, "state": state}
    if scan is not None:
        status["scan"] = scan
    if message is not None:
        status["message"] = message
    return status
